torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,244 @@
+    if (error < 0)  // ignore errors
+      continue;
+    if ((infop.si_code == CLD_EXITED && infop.si_status != 0) ||  // exit with error
+        (infop.si_code == CLD_KILLED) ||
+        (infop.si_code == CLD_DUMPED)) {
+      _exit(EXIT_FAILURE);
+    }
+  }
+
+  // Release ``lock''
+  write(comm_pipe[1], &c, 1);
+
+  // Call the overridden handler.
+  if ((orig_SIGCHLD_sa.sa_flags | SA_SIGINFO) != 0) {
+    // handler is sa_sigaction, this shouldn't really be SIG_IGN or SIG_DFL, but
+    // sa_sigaction and sa_handler happen to be a union, and this fact is
+    // apparently used by Python, so check here.
+    // https://stackoverflow.com/a/24080440
+    if (orig_SIGCHLD_sa.sa_sigaction == (void (*)(int, siginfo_t *, void *)) SIG_IGN) {
+      // SIG_IGN for SIGCHLD is to reap the child and do nothing else.
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_sigaction != (void (*)(int, siginfo_t *, void *)) SIG_DFL) {
+      // SIG_DFL for SIGCHLD is to leave the child as a zombie (do nothing)
+      orig_SIGCHLD_sa.sa_sigaction(sig, info, ctx);
+    }
+  } else {
+    // handler is sa_handler
+    if (orig_SIGCHLD_sa.sa_handler == SIG_IGN) {
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_handler != SIG_DFL) {
+      orig_SIGCHLD_sa.sa_handler(sig);
+    }
+  }
+}
+
+static int isSIGCHLDHanderSet() {
+  struct sigaction sa;
+  int error = sigaction(SIGCHLD, NULL, &sa);
+  if (error == 0) {
+    return ((sa.sa_flags | SA_SIGINFO) != 0) && (sa.sa_sigaction == &handler_SIGCHLD_main);
+  } else {
+    throw std::runtime_error("An error occurred while checking DataLoader SIGCHLD handler");
+  }
+}
+
+// We don't want to exit on any SIGCHLD from any child. child_pids is a tuple
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  HANDLE_TH_ERRORS
+  // assert these types are lock free, just to be safe
+  THPUtils_assert(worker_pids.is_lock_free(), "worker_pids is not lock free");
+  THPUtils_assert(num_worker_pids.is_lock_free(), "num_worker_pids is not lock free");
+
+  THPUtils_assert(PyTuple_Check(child_pids), "_set_main_signal_handlers_for_workers "
+        "expects a tuple, but got %s", THPUtils_typename(child_pids));
+
+  if (comm_pipe[0] == -1) {
+    // we have GIL here so we are fine
+    if (pipe(comm_pipe) != 0) {
+      throw std::runtime_error("An error occurred while setting DataLoader SIGCHLD handler");
+    }
+    char c = '_';
+    write(comm_pipe[1], &c, 1);
+  }
+
+  auto size = PyTuple_GET_SIZE(child_pids);
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = PyTuple_GET_ITEM(child_pids, idx);
+    worker_pid_set.insert((pid_t) THPUtils_unpackLong(obj));
+  }
+  updatePIDsArray();
+
+  // To avoid chain calling our handler, check if the current handler is already
+  // set as ours.
+  if (!isSIGCHLDHanderSet()) {
+    setSignalHandler(SIGCHLD, &handler_SIGCHLD_main, &orig_SIGCHLD_sa);
```

Comment:
We should have a global lock for adding / removing workers, or we can race with setting / removing handlers